File: commands/server_commands/updateNicknames.js
const { getMinecraftNameByUUID } = require('../../other_functions/helperFunctions.js');
const { getAllWhitelistData } = require('../../other_functions/whitelistDatabaseFuncs.js');
const { SlashCommandBuilder } = require('discord.js');

module.exports = {
    data: new SlashCommandBuilder()
        .setName('update_nicknames')
        .setDescription('update nicknames'),
    async execute(client, interaction) {
        if(interaction.user.id !== "336891653291900949") {
            return interaction.reply({
                content: 'You do not have permission to run this command',
            });
        }

        await interaction.deferReply({ ephemeral: true });
        try {
            const whitelist = await getAllWhitelistData();
            const guild = interaction.guild;
            
            for (const element of whitelist) {
                try {
                    const member = await guild.members.fetch(element.discordID).catch(() => null);
                    await interaction.editReply(`Updating nickname for ${member.nickname}...`);
                    if (member && member.nickname !== element.username) {
                        await member.setNickname(element.username).catch(err => {
                            console.error(`Could not set nickname for ${element.discordID}: ${err.message}`);
                        });
                    }
                } catch (error) {
                    console.error(`Error updating nickname for ${element.discordID}:`, error.message);
                }
            }

            await interaction.editReply('Nicknames updated.');
            
        } catch (err) {
            console.log(err);
            await interaction.editReply('An error occurred while updating nicknames.');
        }
    }
};

File: src/main/java/com/createciv/discord_bot/commands/server/whitelist/UpdateNicknames.java
package com.createciv.discord_bot.commands.server.whitelist;

import com.createciv.discord_bot.Bot;
import com.createciv.discord_bot.ConfigLoader;
import com.createciv.discord_bot.classes.SlashCommand;
import com.createciv.discord_bot.util.database.DatabaseRegistry;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;

import java.sql.SQLException;
import java.util.List;
import java.util.Objects;

public class UpdateNicknames extends SlashCommand {

    public UpdateNicknames() {
        super("update_nicknames", "Update nicknames of all discord members");
    }

    @Override
    public void execute(SlashCommandInteractionEvent interactionEvent) {
        if (!hasPermission(interactionEvent)) return;
        List<WhitelistEntry> whitelistEntries;
        WhitelistManager whitelistManager  = (WhitelistManager) DatabaseRegistry.getTableManager("whitelist");
        try {
            whitelistEntries = whitelistManager.getAll();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        Guild guild = interactionEvent.getGuild();
        List<Member> members = Objects.requireNonNull(interactionEvent.getGuild()).getMembers();

        for (WhitelistEntry entry : whitelistEntries) {
            if (members.stream().toList().contains(guild.getMemberById(entry.discordID))) {
                Member member = guild.getMemberById(entry.discordID);
                member.modifyNickname(entry.username);
            }
        }


    }


    private boolean hasPermission(SlashCommandInteractionEvent interactionEvent) {

        if (ConfigLoader.ADMIN_ROLE_IDS.isEmpty()) {
            Bot.LOGGER.error("No admin roles selected. Please set this in the config to continue");
            interactionEvent.reply("Admin roles are not configured. Please configure them to use this command")
                    .setEphemeral(true).queue();
            return false;
        }

        if (interactionEvent.getMember() == null) {
            interactionEvent.reply("Could not find member information").setEphemeral(true).queue();
            return false;
        }

        boolean hasAdminRole = interactionEvent.getMember().getRoles().stream()
                .map(Role::getId)
                .anyMatch(ConfigLoader.ADMIN_ROLE_IDS::contains);

        if (!hasAdminRole) {
            interactionEvent.reply("You do not have permission to run this command").setEphemeral(true).queue();
            return false;
        }
        return true;
    }
}
